lib/net.c

```diff
@@ -700,6 +694,141 @@ get_if_back(struct net_config *net_conf)
 	return net_conf->back_iface_enabled ? &net_conf->back : NULL;
 }
 
+/*
+ * It seems common that devices do not exactly support the
+ * hashes in the DPDK macros such as ETH_RSS_IP, so until we
+ * choose set of minimum hash functions required (instead of
+ * ETH_RSS_IP which is overkill), issue a warning in the
+ * case that the device does not support the hashes.
+ *
+ * TODO #152 Find the minimum set of hash functions (ETH_RSS_*) that
+ * Gatekeeper needs and set
+ * gatekeeper_template_port_conf.rx_adv_conf.rss_conf.rss_hf accordingly.
+ * Then, change this warning to an error and return -1.
+ */
+#define GATEKEEPER_RSS_HF (ETH_RSS_IP)
+
+static int
+check_port_rss(struct gatekeeper_if *iface, uint16_t port_id,
+	struct rte_eth_dev_info *dev_info, struct rte_eth_conf *port_conf)
+{
+	if (dev_info->flow_type_rss_offloads == 0) {
+		/* This port doesn't support RSS, so disable RSS. */
+		G_LOG(NOTICE, "net: port %hu does not support RSS\n", port_id);
+		iface->rss = false;
+	} else if ((dev_info->flow_type_rss_offloads & GATEKEEPER_RSS_HF) !=
+			GATEKEEPER_RSS_HF) {
+		G_LOG(WARNING,
+			"net: port %hu only supports RSS hash functions 0x%"PRIx64", but Gatekeeper requires: 0x%"PRIx64"\n",
+			port_id, dev_info->flow_type_rss_offloads,
+			(uint64_t)GATEKEEPER_RSS_HF);
+	}
+
+	port_conf->rx_adv_conf.rss_conf.rss_hf &=
+		dev_info->flow_type_rss_offloads;
+
+	return 0;
+}
+
+static int
+check_port_mtu(struct gatekeeper_if *iface, uint16_t port_id,
+	struct rte_eth_dev_info *dev_info, struct rte_eth_conf *port_conf)
+{
+	if (dev_info->max_rx_pktlen < iface->mtu) {
```

Comment:
Style: we should replace `iface->mtu` with `port_conf->rxmode.max_rx_pkt_len`, so `check_port_XXX()` functions only compare `dev_info` against `port_conf`.